aggregator/handler/sharded.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3aggregator/aggregator/handler/common"
+	"github.com/m3db/m3aggregator/aggregator/handler/writer"
+	"github.com/m3db/m3aggregator/sharding"
+
+	"github.com/uber-go/tally"
+)
+
+// SharderRouter contains a sharder id and a router.
+type SharderRouter struct {
+	sharding.SharderID
+	common.Router
+}
+
+type sharderRouters struct {
+	SharderID sharding.SharderID
+	Routers   []common.Router
+}
+
+type shardedHandler struct {
+	routersBySharderID []SharderRouter
+	writerOpts         writer.Options
+}
+
+func newShardedHandler(srs []SharderRouter, writerOpts writer.Options) aggregator.Handler {
```

Comment:
nit: is it worth making this constructor public as well since the other handler constructors are public, for example `NewLoggingHandler` and `NewBlackholeHandler`?

Reply:
I'd actually make all of them private will do.